File: src/valueindex/backtest.py
"""Toy backtester for the short-term lab page.

Signals are computed on close prices; entries apply to the NEXT day's
return (no look-ahead). Pure functions, unit-testable.
"""
from __future__ import annotations

from dataclasses import dataclass

import numpy as np
import pandas as pd


def rsi(close: pd.Series, window: int = 14) -> pd.Series:
    delta = close.diff()
    gain = delta.clip(lower=0).rolling(window).mean()
    loss = (-delta.clip(upper=0)).rolling(window).mean()
    rs = gain / loss.replace(0, np.nan)
    return 100 - 100 / (1 + rs)


def rule_rsi_oversold(ohlc: pd.DataFrame, threshold: float = 30, hold: int = 5) -> pd.Series:
    """Long for `hold` days after RSI(14) closes below threshold."""
    signal = rsi(ohlc["close"]) < threshold
    return signal.rolling(hold, min_periods=1).max().fillna(0).astype(float)

def rule_golden_cross(ohlc: pd.DataFrame, fast: int = 50, slow: int = 200) -> pd.Series:
    """Long while the fast moving average is above the slow one."""
    close = ohlc["close"]
    return (close.rolling(fast).mean() > close.rolling(slow).mean()).astype(float)


def rule_down_streak(ohlc: pd.DataFrame, streak: int = 3, hold: int = 5) -> pd.Series:
    """Long for `hold` days after `streak` consecutive down closes."""
    down = ohlc["close"].diff() < 0
    run = down.rolling(streak).sum() >= streak
    return run.rolling(hold, min_periods=1).max().fillna(0).astype(float)


def rule_dip_buy(ohlc: pd.DataFrame, dip_pct: float = 10, hold: int = 20) -> pd.Series:
    """Long for `hold` days when price is `dip_pct`% below its 1y high."""
    close = ohlc["close"]
    high_1y = close.rolling(252, min_periods=60).max()
    signal = close <= high_1y * (1 - dip_pct / 100)
    return signal.rolling(hold, min_periods=1).max().fillna(0).astype(float)


def rule_monday(ohlc: pd.DataFrame) -> pd.Series:
    """Long only on Mondays (weekend-effect test)."""
    idx = pd.DatetimeIndex(ohlc["date"]) if "date" in ohlc else ohlc.index
    return pd.Series((idx.dayofweek == 0).astype(float), index=ohlc.index)


RULES = {
    "rsi": ("RSI 30 이하 매수 (5일 보유)", rule_rsi_oversold),
    "golden_cross": ("골든크로스 (50일 > 200일 이평선)", rule_golden_cross),
    "down_streak": ("3일 연속 하락 후 매수 (5일 보유)", rule_down_streak),
    "dip_buy": ("고점 대비 -10% 진입 (20일 보유)", rule_dip_buy),
    "monday": ("월요일 효과 (월요일만 보유)", rule_monday),
}


@dataclass(frozen=True)
class BacktestResult:
    n_trades: int            # number of entry events (0 -> 1 transitions)
    n_days_in_market: int
    win_rate: float          # share of in-market days with positive return
    mean_daily_ret: float    # average daily return while in market, percent
    t_stat: float            # vs zero mean of in-market daily returns
    total_costs_pct: float   # cumulative cost drag, percent
    equity_curve: pd.Series  # strategy cumulative growth (starts at 1)
    buy_hold_curve: pd.Series


def evaluate(positions: pd.Series, close: pd.Series, cost_bps: float = 0.0) -> BacktestResult:
    """Apply positions to next-day returns with per-side transaction costs."""
    positions = positions.fillna(0).clip(0, 1)
    daily_ret = close.pct_change().fillna(0)
    # Signal on day t earns day t+1's return: shift positions forward.
    held = positions.shift(1).fillna(0)
    strat_ret = held * daily_ret
    # Cost charged on every position change (entry and exit).
    turns = positions.diff().abs().fillna(positions.iloc[0])
    cost = turns * (cost_bps / 10_000)
    strat_ret_net = strat_ret - cost

    in_market = strat_ret[held > 0]
    n_days = len(in_market)
    if n_days > 1 and in_market.std() > 0:
        t_stat = in_market.mean() / (in_market.std() / np.sqrt(n_days))
    else:
        t_stat = 0.0
    entries = int(((positions > 0) & (positions.shift(1).fillna(0) == 0)).sum())

    return BacktestResult(
        n_trades=entries,
        n_days_in_market=n_days,
        win_rate=float((in_market > 0).mean()) if n_days else 0.0,
        mean_daily_ret=float(in_market.mean() * 100) if n_days else 0.0,
        t_stat=float(t_stat),
        total_costs_pct=float(cost.sum() * 100),
        equity_curve=(1 + strat_ret_net).cumprod(),
        buy_hold_curve=(1 + daily_ret).cumprod(),
    )

File: docs/assets/js/backtest.js
// Port of valueindex/backtest.py with pandas semantics preserved:
// full-window rolling means (leading NaN), SMA-variant RSI, signal on
// close applied to next day's return, cost on |position change|.

import { mean, std } from "./stats.js";

const NaNArr = (n) => new Array(n).fill(NaN);

function rollingMean(values, window, minPeriods = window) {
  const out = NaNArr(values.length);
  let sum = 0, count = 0;
  const q = [];
  for (let i = 0; i < values.length; i++) {
    const v = values[i];
    q.push(v);
    if (!Number.isNaN(v)) { sum += v; count++; }
    if (q.length > window) {
      const old = q.shift();
      if (!Number.isNaN(old)) { sum -= old; count--; }
    }
    const inWin = Math.min(i + 1, window);
    if (count >= minPeriods && count === inWin) out[i] = sum / count;
    else if (count >= minPeriods && minPeriods < window) out[i] = sum / count;
  }
  return out;
}

function rollingMax(values, window, minPeriods = window) {
  const out = NaNArr(values.length);
  for (let i = 0; i < values.length; i++) {
    const lo = Math.max(0, i - window + 1);
    let m = -Infinity, count = 0;
    for (let j = lo; j <= i; j++) {
      if (!Number.isNaN(values[j])) { m = Math.max(m, values[j]); count++; }
    }
    if (count >= minPeriods) out[i] = m;
  }
  return out;
}

export function rsi(close, window = 14) {
  const n = close.length;
  const gain = NaNArr(n), loss = NaNArr(n);
  for (let i = 1; i < n; i++) {
    const d = close[i] - close[i - 1];
    gain[i] = Math.max(d, 0);
    loss[i] = Math.max(-d, 0);
  }
  const g = rollingMean(gain, window), l = rollingMean(loss, window);
  return g.map((gi, i) => {
    const li = l[i];
    if (Number.isNaN(gi) || Number.isNaN(li) || li === 0) return NaN;
    return 100 - 100 / (1 + gi / li);
  });
}

function holdAfterSignal(signal, hold) {
  // rolling(hold, min_periods=1).max() over 0/1
  const out = new Array(signal.length).fill(0);
  let lastTrue = -Infinity;
  for (let i = 0; i < signal.length; i++) {
    if (signal[i]) lastTrue = i;
    out[i] = i - lastTrue < hold ? 1 : 0;
  }
  return out;
}

export const RULES = {
  rsi: {
    label: "RSI 30 이하 매수 (5일 보유)",
    fn: (close) => holdAfterSignal(rsi(close).map((r) => r < 30), 5),
  },
  golden_cross: {
    label: "골든크로스 (50일 > 200일 이평선)",
    fn: (close) => {
      const fast = rollingMean(close, 50), slow = rollingMean(close, 200);
      return fast.map((f, i) => (f > slow[i] ? 1 : 0)); // NaN compare -> false
    },
  },
  down_streak: {
    label: "3일 연속 하락 후 매수 (5일 보유)",
    fn: (close) => {
      const down = close.map((c, i) => (i > 0 && c - close[i - 1] < 0 ? 1 : 0));
      const runs = rollingMean(down, 3).map((m) => !Number.isNaN(m) && m * 3 >= 3);
      return holdAfterSignal(runs, 5);
    },
  },
  dip_buy: {
    label: "고점 대비 -10% 진입 (20일 보유)",
    fn: (close) => {
      const high = rollingMax(close, 252, 60);
      const sig = close.map((c, i) =>
        !Number.isNaN(high[i]) && c <= high[i] * 0.9);
      return holdAfterSignal(sig, 20);
    },
  },
  monday: {
    label: "월요일 효과 (월요일만 보유)",
    fn: (close, dates) =>
      dates.map((d) => ((new Date(d + "T00:00:00Z").getUTCDay() + 6) % 7 === 0 ? 1 : 0)),
  },
};

export function evaluate(positions, close, costBps = 0) {
  const n = close.length;
  const dailyRet = new Array(n).fill(0);
  for (let i = 1; i < n; i++) dailyRet[i] = close[i] / close[i - 1] - 1;

  const held = [0, ...positions.slice(0, n - 1)];
  const strat = held.map((h, i) => h * dailyRet[i]);
  const turns = positions.map((p, i) => (i === 0 ? Math.abs(p) : Math.abs(p - positions[i - 1])));
  const costRate = costBps / 10000;
  let totalCost = 0, equity = 1, buyHold = 1;
  const equityCurve = new Array(n), buyHoldCurve = new Array(n);
  for (let i = 0; i < n; i++) {
    const cost = turns[i] * costRate;
    totalCost += cost;
    equity *= 1 + (strat[i] - cost);
    buyHold *= 1 + dailyRet[i];
    equityCurve[i] = equity;
    buyHoldCurve[i] = buyHold;
  }

  const inMarket = strat.filter((_, i) => held[i] > 0);
  const nDays = inMarket.length;
  let tStat = 0;
  if (nDays > 1) {
    const s = std(inMarket);
    if (s > 0) tStat = mean(inMarket) / (s / Math.sqrt(nDays));
  }
  let entries = 0;
  for (let i = 0; i < n; i++) {
    const prev = i === 0 ? 0 : positions[i - 1];
    if (positions[i] > 0 && prev === 0) entries++;
  }
  return {
    n_trades: entries,
    n_days_in_market: nDays,
    win_rate: nDays ? inMarket.filter((x) => x > 0).length / nDays : 0,
    mean_daily_ret: nDays ? mean(inMarket) * 100 : 0,
    t_stat: tStat,
    total_costs_pct: totalCost * 100,
    equityCurve,
    buyHoldCurve,
  };
}
